common/predictive-text/unit_tests/headless/worker-model-compositor.js

```diff
@@ -50,6 +50,45 @@ describe('ModelCompositor', function() {
         });
       });
 
+      it('strongly avoids corrections for single-character roots', function() {
+        let compositor = new ModelCompositor(plainModel);
+        let context = {
+          left: '', startOfBuffer: true, endOfBuffer: true,
+        };
+
+        // The 'weights' involved imply that we have an edge-case fat finger on the bottom of
+        // the 'q' key, slightly in its favor.
+        let inputDistribution = [
+          {sample: {insert: 'q', deleteLeft: 0}, p: 0.5},  // 'quite' (679) and 'question' (644) are included!
+          {sample: {insert: 'a', deleteLeft: 0}, p: 0.4}   // but at lower weight than 'and' (998).
+        ];
+
+        compositor.predict({insert: '', deleteLeft: 0}, context); // Initialize context tracking first!
+        let suggestions = compositor.predict(inputDistribution, context);
+
+        // remove the keep suggestion; we're not testing that here.
+        suggestions = suggestions.filter((suggestion) => suggestion.tag != 'keep');
+
+        suggestions.sort((a, b) => b.p - a.p);
+
+        // There are only 4 suggestions in this limited test model that begin with 'q'.
+        // We expect more than that, since 'a' is indicated to be very close by.
+        assert.isAbove(suggestions.length, 4, "fat-finger style corrections needed for test comparisons are missing");
+
+        // Note:  'and' is (currently) modeled by the text-fixture model to have 9.3x the base probability
+        // that the worst 'q'-rooted suggestion ('quality') does.  Without single-character correction
+        // avoidance logic, this test _will_ fail.
+        //
+        // In case a tweak to test parameters is desired, note that 'and' beats rank #3 - 'questions' -
+        // at 3.36x base.  At the time of writing this test, upping 'a's probability to 0.45 will block
+        // 'quality' while the top three 'q's (ending with 'questions') remain in place.
```

Comment:
With the parameterization I chose, putting 'a' to 0.45 yields (0.5 / 0.45)^16 = (1.1111)^16 = 5.3966.  'quality' needs a stronger boost than this.  Roughly speaking, just doubling the fractional part of the ratio is enough to square the boost!

(When they're close to even; this does drop off somewhat when the ratio is more pronounced.)